src/Symfony/Component/HttpClient/Retry/HttpStatusCodeDecider.php

```diff
@@ -11,25 +11,56 @@
 
 namespace Symfony\Component\HttpClient\Retry;
 
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+
 /**
  * Decides to retry the request when HTTP status codes belong to the given list of codes.
  *
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
 final class HttpStatusCodeDecider implements RetryDeciderInterface
 {
+    public const METHOD_DECISION = [
+        'GET' => true,
+        'HEAD' => true,
+        'PUT' => true,
+        'DELETE' => true,
+        'OPTIONS' => true,
+        'TRACE' => true,
+    ];
+    public const STATUS_CODES_DECISION = [
+        423 => true,
+        425 => true,
+        429 => true,
+        500 => self::METHOD_DECISION,
+        502 => true,
+        503 => true,
+        504 => self::METHOD_DECISION,
+        507 => self::METHOD_DECISION,
+        510 => self::METHOD_DECISION,
+    ];
+
     private $statusCodes;
 
     /**
-     * @param array $statusCodes List of HTTP status codes that trigger a retry
+     * @param array $statusCodes List of HTTP status codes that trigger a retry. Either a boolean to retry/not-retry or a list of request method to take into account
      */
-    public function __construct(array $statusCodes = [423, 425, 429, 500, 502, 503, 504, 507, 510])
+    public function __construct(array $statusCodes = self::STATUS_CODES_DECISION)
     {
         $this->statusCodes = $statusCodes;
     }
 
-    public function shouldRetry(string $requestMethod, string $requestUrl, array $requestOptions, int $responseStatusCode, array $responseHeaders, ?string $responseContent): ?bool
+    public function shouldRetry(int $retryCount, string $requestMethod, string $requestUrl, array $requestOptions, int $responseStatusCode, array $responseHeaders, ?string $responseContent, ?TransportExceptionInterface $exception): ?bool
     {
-        return \in_array($responseStatusCode, $this->statusCodes, true);
+        if ($exception !== null && !isset($this->statusCodes[$responseStatusCode])) {
```

Comment:
So, if there is an exception AND if the status code is not in the list of things we should retry... then rewrite the status code?

I don't understand why we rewrite the status code.

Reply:
This code is: If there is an exception AND the status code is not defined, then we treat the exception like a 500 error..

I don't really like it either, but I don't had better idea...

Reply:
hm.. what's about:
```php
$statusCodes[
  423 => true,
  500 => self::METHOD_DECISION,
  TransportExceptionInterface::class => self::METHOD_DECISION,
];
```
/cc @nicolas-grekas